foodji_api/Application/Command/UpdateRecipeCommand.cs

```diff
@@ -0,0 +1,62 @@
+﻿using Application.Dto;
+using AutoMapper;
+using Infra;
+using MediatR;
+using Domain.Recipes;
+using MongoDB.Bson;
+using MongoDB.Driver;
+
+namespace Application.Command;
+
+public class UpdateRecipeCommand : IRequest<string>
+{
+    private RecipeDto RecipeDto { get; }
+    
+    public UpdateRecipeCommand(RecipeDto recipeDto)
+    {
+        RecipeDto = recipeDto;
+    }
+    
+    private class Handler : IRequestHandler<UpdateRecipeCommand, string>
+    {
+        private readonly IFoodjiDbClient _client;
+        private readonly IMapper _mapper;
+
+        public Handler(IFoodjiDbClient client, IMapper mapper)
+        {
+            _client = client;
+            _mapper = mapper;
+        }
+        
+        public async Task<string> Handle(UpdateRecipeCommand request, CancellationToken cancellationToken)
+        {
+            var recipe = _mapper.Map<Recipe>(request.RecipeDto);
+            
+            var results = await _client.Recipes.FindAsync(
+                x => x.Id == new ObjectId(request.RecipeDto.Id),
+                cancellationToken: cancellationToken);
+
+            var recipeToUpdateList = results.ToList(cancellationToken);
+            
+            switch (recipeToUpdateList.Count)
+            {
+                case > 1:
+                    // TODO More specific exception to go along better exception handling in the API layer
+                    //      500 many with the same ID (bad news!)
+                    //      shouldn't happen, but no "FindOne" method to make that check for us
+                    throw new Exception($"{recipeToUpdateList.Count} recipes with the id {request.RecipeDto.Id} found");
+                case 0:
+                    return string.Empty;
+            }
+
+            var recipeToUpdate = recipeToUpdateList.First();
+            
+            recipeToUpdate.Update(recipe);
+            
+            await _client.Recipes.ReplaceOneAsync(r => r.Id.Equals(recipeToUpdate.Id),
```

Comment:
Je me demande si ça vaut la peine de faire un *autre* check ici que la recette existe - ça reste deux requêtes séparées, async de surcroit, et un `DELETE` *pourrait* se nicher entre les deux requêtes Mongo. Je me demande aussi si ce petit check ne pourrait pas faire parti d'une classe utilitaire dans le projet Infra - Une méthode templatée du genre

```cs
public T? GetSingleQueryResult(IEnumerable<T> result)
{
    if (result.Count > 1) 
    {
        // Moins spécifique ici sur le message d'erreur, admittedly, mais je penses qu'un message d'erreur
        // pour un nombre de résultats >1 dans une query est raisonable, 
        // surtout que ce n'est vraiment pas un comportement auquel on s'attend
        throw new Exception("Yo your DB is fucked my guy, got more than one result on a unique search, what's up with that?");
    }
    return result.Count == 0 ? null : result.First();
}
```

C'est un peu coin de table, et si on veut extraire l'id et le caster en string, on doit quand même faire un null-check à la sortie de la méthode, mais si on veut juste l'objet, ça simplifie le code à la sortie. T'en penses quoi?

Reply:
Premièrement, dans ce cas-là, je crois que renvoyer l'id est plenty enough, surtout que l'objet est legit supprimé donc inutile pour le frontend.

On pourrait en effet faire un premier find pour trouver une (pas zéro, pas deux) recette à supprimer et ensuite caller le delete dessus. Cependant, il faudrait valider (un peu comme le update, en fait?) qu'il est impossible que deux transactions se fassent en même temps et impactent le même objet/id. PAR CONTRE, croire (ou douter) qu'il serait possible d'avoir 2 fois le même id dans une même collection, c'est carrément perdre foi en mongodb car c'est lui qui génère les id et même si on lui force la main avec un id prégénéré, il nous recrache un 500 et nous dit git gud si on a un conflit. De surcroît, si jamais un cas comme ça se passe, je suis d'avis que supprimer les deux ressources est probablement la bonne décision anyway, ça fait un cleanup 😛. Si on veut y aller avec ça, il faudrait changer le delete pour un DeleteManyAsync, mais je ne crois pas que c'est nécessaire, le DeleteOneAsync fait la job.

Finalement, si on ping 0 recettes, le deleteOneAsync gère le cas pépère et retourne un null, ce qu'on valide déjà et on retourne un 404 not found. Dans ce cas, faire un find avant ne change absolument rien.

En bref, je ne crois pas que de valider le count pour un delete soit nécessaire. Je crois par contre que d'extraire le bloc de code pour valider le count pourrait être pertinent et dans la méthode on aurait basically juste le switchcase qui throw l'exception si plus qu'une recette. La méthode pourrait soit retourner l'élément (comme dans ton exemple) ou retourner un bool qui valide le count de la liste.

Reply:
Très d'accord pour le fait que vérifier le Count sur le Delete est redondant.

Pour ce qui est de la vérification, résolu on call par LINQ. Vive `SingleOrDefault()`!